ci: configure semantic-release pipeline

antongolub opened the pull request:
closes #64

Adds `semantic-release` action.
env config example:
```
      GITHUB_TOKEN: 'token'
      GH_TOKEN: 'token'
      NPM_TOKEN: 'token'
      GIT_AUTHOR_EMAIL: 'mailbox@antongolub.ru'
      GIT_COMMITTER_EMAIL: 'mailbox@antongolub.ru'
      GIT_AUTHOR_NAME: '@antongolub'
      GIT_COMMITTER_NAME: '@antongolub'
```

antongolub commented:
@FarazzShaikh, could you also take a look?

antonmedv commented:
I not like to rely third party actions.

FarazzShaikh commented:
In that case, how about using [GitHub CLI](https://cli.github.com/manual/gh_release_create)  to create a release with bash commands?

antongolub commented:
NP, `semrel` may be called through `npx`

FarazzShaikh commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,47 @@
+  release:
+    name: Release
+    # https://github.community/t/trigger-job-on-tag-push-only/18076
+    if: github.event_name == 'push' && github.ref == 'refs/heads/master'
```
Why not use:
```yaml
on:
  push:
    branches:
      - master
```
I assume you're trying to limit this to pushes to the master branch correct?

antongolub replied:
This is just a preparation for extended push-pull flow. Anyway, fixed for now.

antonmedv commented:
Canxwe avoid any deps for releasing?

antongolub commented:
If we drop `semrel`, there will be no changelog, auto-versioning, so on.

antonmedv commented:
Closing as already implemented.

antongolub commented:
Not really: no changelog, no conventional commits.
If any third-party solutions are not allowed, I suggest replacing `semrel` with... zx-script )). This would be a nice practical demo. 

```javascript
// Replaces semantic-release with zx script

const semanticTagPattern =  /^(\d+)\.(\d+)\.(\d+)$/
const releaseSeverityOrder = ['major', 'minor', 'patch']
const semanticRules = [
  {group: 'Features', releaseType: 'minor', prefixes: ['feat', 'Update']},
  {group: 'Fixes & improvements', releaseType: 'patch', prefixes: ['fix', 'perf', 'refactor']},
  {group: 'BREAKING CHANGES', releaseType: 'major', keywords: ['BREAKING CHANGE', 'BREAKING CHANGES']},
]

const tags = (await $`git tag --merged`).toString().split('\n')
const lastVersion = tags.filter(tag => semanticTagPattern.test(tag)).pop()
const newCommits = (lastVersion
  ? await $`git log --format=%s___%b___%h___%H ${await $`git rev-list -1 ${lastVersion}`}..HEAD`
  : await $`git log --format=%s___%b___%h___%H HEAD`)
    .toString().trim()
    .split('\n')
    .map(msg => {
      const [subj, body, short, hash] = msg.split('___')
      return {subj, body, short, hash}
    })

const semanticChanges = newCommits.reduce((acc, {subj, body, short, hash}) => {
  semanticRules.forEach(({group, releaseType, prefixes, keywords}) => {
    const prefixMatcher = prefixes && new RegExp(`^(${prefixes.join('|')})(\\(\w+\\))?\\s.+$`)
    const keywordsMatcher = keywords && new RegExp(`(${keywords.join('|')})`)

    if (prefixMatcher && prefixMatcher.test(subj) || keywordsMatcher && keywordsMatcher.test(body)) {
      acc.push({
        group,
        releaseType,
        subj,
        body,
        short,
        hash
      })
    }
  })

  return acc
}, [])

const resolveNextVersion = (lastVersion, releaseType) => {
  if (!releaseType) {
    return
  }
  if (!lastVersion) {
    return '1.0.0'
  }

  const [,c1, c2, c3] = semanticTagPattern.exec(lastVersion)

  if (releaseType === 'major') {
    return `${1 + +c1}.0.0`
  }

  if (releaseType === 'minor') {
    return `${c1}.${1 + +c2}.0`
  }

  if (releaseType === 'patch') {
    return `${c1}.${c2}.${1 + +c3}`
  }

}
const nextReleaseType = releaseSeverityOrder.find(type => semanticChanges.find(({releaseType}) => type === releaseType))
const nextVersion = resolveNextVersion(lastVersion, nextReleaseType)

console.log('semanticChanges=', semanticChanges)
console.log('lastVersion=', lastVersion)
console.log('nextReleaseType=', nextReleaseType)
console.log('nextVersion=', nextVersion)

// const releaseCommit = semanticChanges.reduce((acc, {group}) => {}, {})

// Next steps:
// 1. update package.json
// 2. create changelog.md or just append new changes
// 3. git add, git commit, git push with tag
// 4. npm publish
// 5. gh release
```

antongolub commented:
@antonmedv, does it make sense to continue?

antonmedv commented:
Yes, it’s better)) but there also problem with git commit from ci due to cla restrictions. So it will take much more effort what maybe not needed.

antongolub commented:
@antonmedv, the trampoline is working: https://github.com/antongolub/zx-semrel

I've done some experiments:
1) GitHub PAT (which grants access to all repos) may be replaced with ssh deploy token attached to the single repo: https://github.com/antongolub/zx-semrel/blob/6148aedb1ad433d9a19b1ef14e775ff7662d838c/.github/workflows/release.yml
2. Unfortunately, ssh can not be used to interact with GitHub API for now.
```
ssh -T git@github.com
Hi antongolub! You've successfully authenticated, but GitHub does not provide shell access.
```
So we have to use a regular oauth token for GitHub release creation.

3. GITHUB_TOKEN may be used as an alternative for user/bot account's PAT. It's dynamic, repo special and [could be emitted with limited permissions](https://github.blog/changelog/2021-04-20-github-actions-control-permissions-for-github_token/) for `gh-actions` workflows.

antonmedv commented:
This is cool. 😎 
But I still not convinced what we need this complication for zx.

antongolub commented:
np. ci-cd should appear when it saves more time than it takes to maintain it )

Dunky13 commented:
Using `semrel` would be really beneficial.
I'm currently writing code that will, hopefully, go into production (soon) :) 
Personally I really love the quick releases, with every new feature. But sometimes it's hard to understand what's new with the release.

I understand that writing a changelog it too tedious, so `semrel` could offload that a bit, and give us (the users) a better overview of what has changed this release. And whether or not to change the code base to the new feature (and discard, if developed, the workaround)

All in all, thank you for your hard work! Love this project :)
